### PR TITLE
Add support for PS4 multipart updates

### PR DESCRIPTION
**Pull request by sarpt**
This PR adds basic support for fetching PS4 multipart updates.
The PS4 update packages come in two variants: multipart full update containing every version to date in parts split below ~4GB, and delta updates which only contain changes since the previous version. This implementation adds support for the former, by fetching information about the latest version and providing list of parts to download:
image

Since I'm not sure which endpoint exposes older versions of PS4 updates (I could only find urls pointing to the latest version), providing an option to download delta update seems to be of lesser priority, since a delta is only useful with older versions. If there's demand for it, or if there's a way to fetch older versions, it could be added later.

Parts are provided to download in a list, just like different versions of PS3 updates are. If the whole update pkg is not split (does not exceed 4GB), the element in a list won't specify a part number. Basically, part entries behave for PS4 updates as different versions entries behave for PS3 updates, both in egui and CLI, and the major difference is that a part number is shown next to the version when it's applicable.

There aren't any changes to file creation behavior and filesystem operations. Hashing is modified to allow hashing of the whole package, since PS4 multipart updates don't seem to embed hash as a suffix to pkg data.

Tested on: Linux 6.6.47-1-MANJARO x86_64 with an Xfce DE. Unfortunately I have neither a Windows nor a Mac machine to test it there. In my case newly added dependencies did not require any additional system packages to be installed.

### CODE REVIEW

**Comment by RainbowCookie32** [comment]
Wow, thanks for the feature! Been meaning to do this for a really long time, but never got around to it. I'll get to reviewing, and I can test on Windows (macOS we'll just hope for the best lol)

File: src/egui/mod.rs [hunk 1]
```diff
@@ -364,7 +367,7 @@ impl UpdatesApp {
             ui.add_enabled_ui(!self.v.serial_query.is_empty() && self.v.search_promise.is_none(), | ui | {
                 if !input_submitted && !ui.button("Search for updates").clicked() { return; }
 
-                let already_searched = self.v.update_results.iter().any(|e| e.title_id == parse_title_id(&self.v.serial_query));
+                let already_searched = self.v.update_results.iter().any(|e| e.title_id == crate::psn::parse_title_id(&self.v.serial_query));
```
Comment (RainbowCookie32) [inline src/egui/mod.rs:370]: Shouldn't `parse_title_id` be available already with the `use::crate::psn::*`?

**Comment by RainbowCookie32** [comment]
By the way, would you mind updating the README to mention that PS4 updates can also be downloaded now?

File: src/egui/mod.rs (hunk 1 above)
Comment (sarpt) [reply src/egui/mod.rs]: Correct, seems that my IDE added this when I was experimenting with modules. Done

File: src/psn/mod.rs [hunk 2]
```diff
@@ -130,7 +168,18 @@ impl PackageInfo {
...
+            hash_whole_file: false,
+            manifest_url: String::new(),
+            offset: 0,
+            part_number: None,
+        }
+    }
+
+    pub fn id(&self) -> String {
+        match self.part_number {
+            Some(part_idx) => format!("{0} - part {1}", self.version, part_idx),
```
[the remark this answers is not in the document]
Comment (sarpt) [reply src/psn/mod.rs]: Done

File: src/psn/utils.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    return None
+}
+
+pub fn get_update_info_url(title_id: &str, platform_variant: PlaformVariant) -> Result<String, UpdateError> {
+    match platform_variant {
+        PlaformVariant::PS3 => {
+            Ok(format!("https://a0.ww.np.dl.playstation.net/tpl/np/{0}/{0}-ver.xml", title_id))
+        },
+        PlaformVariant::PS4 => {
+            let key = match hex::decode("AD62E37F905E06BC19593142281C112CEC0E7EC3E97EFDCAEFCDBAAFA6378D84") {
```
[the remark this answers is not in the document]
Comment (sarpt) [reply src/psn/utils.rs:41]: It's a public key for HMAC encryption, listed here: https://www.psdevwiki.com/ps4/Keys as "HMAC-SHA256 Patch Pkg URL Key". The title id is prefixed with "np_" and hashed as a message using this key, which is then inserted into the url

**Comment by sarpt** [comment]
> By the way, would you mind updating the README to mention that PS4 updates can also be downloaded now?

Done, I've added "PS3 & PS4" to the readme.

**Comment by RainbowCookie32** [comment]
Well, everything seems to be working so I'll be merging and tagging a new release. Thanks again for the feature!

**Pull request merged**